File: WordBreakBF.java
//TC : O(k^n) , k is the length of dictionary
//SC : O(1)
class Solution {
    HashSet<String> set ;
    public boolean wordBreak(String s, List<String> wordDict) {
        set = new HashSet<>(wordDict);

        if(s.length() == 0 || s==null) return false;


        return helper(s);

    }

    private boolean helper(String s){
         if(s.length() == 0) return true;


         for(int i=0;i<=s.length();i++){

             if(set.contains(s.substring(0,i)) && helper(s.substring(i)))
                return true;
         }


        return false;

    }
}

File: wordBreak_BF.py
"""
Bruete Force solution Time limit exceeded

exponential time of len of string

"""


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        self.d=wordDict
        if(s is None or len(s)==0):
            return False
        self.d=wordDict
        return self.helper(s)
    
    
    def helper(self,s):
        #base
        if(len(s)==0):
            return True
        
        #logic
        for i in range(len(s)):
            if s[0:i+1] in self.d and self.helper(s[i+1:]):
                return True
        
        return False
                